File: src/admin/ops.ts
import { backendIds } from '../agent';
import { catalogById } from '../agent/catalog';
import type { AgentBackend, BackendProbe, PermissionMode } from '../agent/types';
import { tierLabel, type BackendProbeRow } from '../card/dm-cards';
import {
  effectiveGuestMode,
  effectiveMode,
  getProjectByName,
  updateProject,
  type Project,
} from '../project/registry';

/**
 * 管理面共享写操作层（第二棒：daemon 集成 + 写操作双端同源）。
 *
 * 【契约】DM 卡片回调（handle-message 的 dm.proj.* / gs.*）与 Web 控制台
 * （AdminService 写方法，经 Orchestrator.adminExecute / supervisor IPC）**都只
 * 走这里**写 projects.json——同一套校验、同一套落盘、同一套活跃会话驱逐，杜绝
 * 双写两套逻辑漂移。本模块只依赖 registry / agent 注册表 / dm-cards 的 tierLabel
 * （纯展示），不依赖 channel / orchestrator——驱逐动作由调用方以回调注入
 * （evictLiveSessionsForChat 是 orchestrator 进程内状态，必须在 bot 进程内执行；
 * 这也是 supervisor 写操作走 IPC 而非文件级直写的根本原因）。
 *
 * 写操作必须在「持有该 bot registry 的进程」内执行：registry 的 withLock 是
 * 进程内锁，跨进程直写会与 bot 进程的读-改-写交错丢更新。
 */

/** Web/IPC 写操作的序列化形态（supervisor → bot 子进程经 process.send 转发）。 */
export type AdminWriteOp =
  | { kind: 'switchBackend'; project: string; backend: string }
  | {
      kind: 'setPermissionMode';
      project: string;
      mode?: PermissionMode;
      guestMode?: PermissionMode;
      network?: boolean;
    }
  | { kind: 'setNoMention'; project: string; on: boolean }
  | { kind: 'setAutoCompact'; project: string; on: boolean };

/** 写操作被校验拒绝（中文原因可直接上卡/上 HTTP body）。HTTP 层映射 409；
 * IPC 层凭 code 还原类型。 */
export class AdminWriteError extends Error {
  readonly code = 'ADMIN_WRITE_REJECTED';
  constructor(reason: string) {
    super(reason);
    this.name = 'AdminWriteError';
  }
}

/** perform* 的统一返回：ok 带写后回读的最新项目（卡片/响应直接渲染它，保证与
 * 盘上一致）；!ok 带中文拒因。不抛错——DM 与 Web 对拒因的呈现方式不同。 */
export type AdminWriteOutcome = { ok: true; project: Project } | { ok: false; reason: string };

const TIERS: readonly PermissionMode[] = ['qa', 'write', 'full'];

/**
 * 项目后端切换的纯校验（exported for tests）：① 目标 id 在注册表里；② 目标后端
 * doctor() 探测通过（未装 CLI / SDK 不可用要在这里拦住，而不是切过去后每条消息
 * 报错）；③ 项目两档权限（管理员档 + 普通用户档）都在目标后端声明的支持面内
 * （某后端若仅支持部分权限档——其 startThread 的 fail-closed 硬守卫不变，这里
 * 只是把拒绝提前到切换时并讲清原因）。返回首个不满足的原因（中文，直接上卡）；
 * 全过返回 null。切换本身不驱逐活跃会话：SessionRecord.backend 让已有话题会话
 * 仍走原后端，新话题才用新值（resolveThread 按记录路由的既有语义）。
 */
export function validateBackendSwitch(opts: {
  /** 目标后端 id（下拉提交值） */
  target: string;
  /** 注册表内全部后端 id（backendIds()） */
  registered: readonly string[];
  /** 项目当前权限档（两档都要被目标后端支持） */
  project: Pick<Project, 'mode' | 'guestMode'>;
  /** 目标后端声明的权限档支持面（AgentBackend.supportedModes；undefined ⇒ 全支持） */
  supportedModes?: readonly PermissionMode[];
  /** 目标后端 doctor() 探测结果；undefined = 探测没跑成，按不可用拒绝 */
  probe?: BackendProbe;
}): string | null {
  if (!opts.registered.includes(opts.target)) {
    return `未知后端「${opts.target}」（可用：${opts.registered.join('、')}）`;
  }
  if (!opts.probe?.ok) {
    return `后端「${opts.target}」当前不可用：${opts.probe?.hint ?? '环境探测失败（未安装或未登录）'}`;
  }
  if (opts.supportedModes) {
    const tiers = [...new Set([effectiveMode(opts.project), effectiveGuestMode(opts.project)])];
    const unsupported = tiers.filter((t) => !opts.supportedModes!.includes(t));
    if (unsupported.length > 0) {
      return (
        `该后端仅支持 ${opts.supportedModes.map(tierLabel).join(' / ')} 权限档，` +
        `本项目当前为 ${tiers.map(tierLabel).join(' / ')} —— 请先在「🔐 权限」把两档都调整到支持的档位再切换。`
      );
    }
  }
  return null;
}

/** 单个后端 doctor 探测的超时兜底（ms）：探测要 spawn CLI，未安装/卡死时不能
 * 拖住检测卡——超时按「探测没跑成」（probe undefined）渲染为不可用。 */
export const BACKEND_PROBE_TIMEOUT_MS = 3000;

/**
 * 🧠 后端检测的纯探测（exported for tests）：并行对全部后端 doctor({force})，
 * 单个超时（Promise.race 兜底）/抛错都归一成 probe undefined——检测结果卡按
 * 不可用渲染，绝不放行。输入是后端实例的最小切面，注册表里有什么测什么，
 * 新后端注册即自动出现在结果卡上。
 */
export async function probeBackends(
  backends: readonly Pick<AgentBackend, 'id' | 'displayName' | 'supportedModes' | 'doctor'>[],
  timeoutMs = BACKEND_PROBE_TIMEOUT_MS,
): Promise<BackendProbeRow[]> {
  return Promise.all(
    backends.map(async (be) => {
      let timer: NodeJS.Timeout | undefined;
      const timeout = new Promise<undefined>((resolve) => {
        timer = setTimeout(() => resolve(undefined), timeoutMs);
      });
      const probe = await Promise.race([be.doctor({ force: true }).catch(() => undefined), timeout]).finally(() =>
        clearTimeout(timer),
      );
      return { id: be.id, name: be.displayName, probe, supportedModes: be.supportedModes };
    }),
  );
}

/** 写后回读（卡片/响应与盘上一致）；理论上 update 后必在，缺了用合成值兜底。 */
async function freshOr(name: string, fallback: Project): Promise<Project> {
  return (await getProjectByName(name)) ?? fallback;
}

/**
 * 🧠 切换项目后端（DM dm.proj.backend.submit 与 Web switchBackend 同源）：
 * 写盘前再 doctor({force}) 探一次「现在」的状态（检测卡渲染后环境可能已变），
 * validateBackendSwitch 全过才写 Project.backend。**不驱逐活跃会话**——已有话题
 * 会话按 SessionRecord.backend 仍走原后端，新话题才用新值。
 */
export async function performBackendSwitch(opts: {
  projectName: string;
  target: string;
  /** 后端实例解析（orchestrator 的缓存版或裸 createBackend 均可） */
  backendFor: (id: string) => AgentBackend;
}): Promise<AdminWriteOutcome> {
  const p = await getProjectByName(opts.projectName);
  if (!p) return { ok: false, reason: `项目「${opts.projectName}」不存在` };
  // 后端在新建项目时选定、运行时固定，不支持切换（防御式：UI 已撤切换入口，这里挡住
  // 任何残留 IPC/HTTP 请求）。仅放行 legacy 项目（backend 未设）的一次性落地与同值 no-op。
  if (p.backend && p.backend !== opts.target) {
    return {
      ok: false,
      reason: '该项目的后端已在创建时选定，运行时固定、不支持切换。如需更改，请删除该项目后用新后端重新创建。',
    };
  }
  const registered = backendIds();
  const known = registered.includes(opts.target);
  const probe = known ? (await probeBackends([opts.backendFor(opts.target)]))[0]?.probe : undefined;
  const reason = validateBackendSwitch({
    target: opts.target,
    registered,
    project: p,
    supportedModes: known ? opts.backendFor(opts.target).supportedModes : undefined,
    probe,
  });
  if (reason) return { ok: false, reason };
  await updateProject(opts.projectName, { backend: opts.target });
  return { ok: true, project: await freshOr(opts.projectName, { ...p, backend: opts.target }) };
}

/**
 * 🔐 设置权限档（DM dm.proj.perm.submit 与 Web setPermissionMode 同源）：落盘
 * 管理员档 mode / 普通用户档 guestMode / 联网，再驱逐本项目活跃会话让新档立即
 * 生效（codex 沙箱在 thread/start 绑定后不可变）。mode/guestMode 缺省 = 不改。
 */
export async function performSetPermissionMode(opts: {
  projectName: string;
  mode?: PermissionMode;
  guestMode?: PermissionMode;
  network?: boolean;
  /** orchestrator 进程内驱逐（LIVE 线程只存在于 bot 进程内存里） */
  evictLiveSessionsForChat: (chatId: string) => Promise<void>;
}): Promise<AdminWriteOutcome> {
  // Web 来路的档位值未经 DM 端 asTier 收窄，这里再守一道（DM 来路恒合法）。
  for (const v of [opts.mode, opts.guestMode]) {
    if (v !== undefined && !TIERS.includes(v)) return { ok: false, reason: `未知权限档「${String(v)}」` };
  }
  const p = await getProjectByName(opts.projectName);
  if (!p) return { ok: false, reason: `项目「${opts.projectName}」不存在` };
  // 后端档位兼容守门：后端在创建时选定、运行时固定（不支持切换），但权限档可改。
  // 若把档改到该后端的 supportedModes 之外（某后端若只支持部分档，却改到档外），
  // 新话题会在 backend.startThread 的 fail-closed 守卫处直接抛错、整群卡死。提前在这里拦住，
  // 给清晰原因，而不是让用户改完才发现聊不了。codex（supportedModes undefined）全档放行。
  const entry = p.backend ? catalogById(p.backend) : undefined;
  if (entry?.supportedModes) {
    const resMode = opts.mode ?? effectiveMode(p);
    const resGuest = opts.guestMode ?? effectiveGuestMode(p);
    const bad = [...new Set([resMode, resGuest])].find((t) => !entry.supportedModes!.includes(t));
    if (bad !== undefined) {
      return {
        ok: false,
        reason:
          `项目的后端「${entry.displayName}」仅支持 ${entry.supportedModes.map(tierLabel).join(' / ')} 权限档，` +
          `无法改到「${tierLabel(bad)}」。该后端是创建时选定、运行时固定的；如需更低权限档，请删项目后改用 codex 后端重建。`,
      };
    }
  }
  // updateProject 跳过 undefined 值 —— 与 DM 旧写法 {...(mode?{mode}:{})} 等价。
  await updateProject(opts.projectName, { mode: opts.mode, guestMode: opts.guestMode, network: opts.network });
  await opts.evictLiveSessionsForChat(p.chatId);
  return { ok: true, project: await freshOr(opts.projectName, p) };
}

/** ✋ 免@ 开关（DM dm.proj.noMention / gs.noMention 与 Web setNoMention 同源）。
 * 即时生效（每条消息读盘判定），无需驱逐。 */
export async function performSetNoMention(opts: { projectName: string; on: boolean }): Promise<AdminWriteOutcome> {
  const p = await getProjectByName(opts.projectName);
  if (!p) return { ok: false, reason: `项目「${opts.projectName}」不存在` };
  await updateProject(opts.projectName, { noMention: opts.on });
  return { ok: true, project: await freshOr(opts.projectName, { ...p, noMention: opts.on }) };
}

/** 🗜️ 自动压缩开关（DM dm.proj.autoCompact / gs.autoCompact 与 Web 同源）：
 * 压缩上限在 thread/start 绑定，落盘后驱逐活跃会话让下一条消息重绑生效。 */
export async function performSetAutoCompact(opts: {
  projectName: string;
  on: boolean;
  evictLiveSessionsForChat: (chatId: string) => Promise<void>;
}): Promise<AdminWriteOutcome> {
  const p = await getProjectByName(opts.projectName);
  if (!p) return { ok: false, reason: `项目「${opts.projectName}」不存在` };
  await updateProject(opts.projectName, { autoCompact: opts.on });
  await opts.evictLiveSessionsForChat(p.chatId);
  return { ok: true, project: await freshOr(opts.projectName, { ...p, autoCompact: opts.on }) };
}

/** Orchestrator.adminExecute 的实现体：AdminWriteOp → 对应 perform*；拒绝抛
 * {@link AdminWriteError}（HTTP 409 / IPC code 还原）。Web/IPC 专用入口——DM
 * 回调直接调 perform* 拿 outcome 渲染卡片，不走这里。 */
export function createAdminWriteExecutor(deps: {
  backendFor: (id?: string) => AgentBackend;
  evictLiveSessionsForChat: (chatId: string) => Promise<void>;
}): (op: AdminWriteOp) => Promise<void> {
  return async (op) => {
    const outcome = await runAdminWriteOp(op, deps);
    if (!outcome.ok) throw new AdminWriteError(outcome.reason);
  };
}

/** AdminWriteOp 分发（exported for tests）。 */
export async function runAdminWriteOp(
  op: AdminWriteOp,
  deps: { backendFor: (id?: string) => AgentBackend; evictLiveSessionsForChat: (chatId: string) => Promise<void> },
): Promise<AdminWriteOutcome> {
  switch (op.kind) {
    case 'switchBackend':
      return performBackendSwitch({ projectName: op.project, target: op.backend, backendFor: deps.backendFor });
    case 'setPermissionMode':
      return performSetPermissionMode({
        projectName: op.project,
        mode: op.mode,
        guestMode: op.guestMode,
        network: op.network,
        evictLiveSessionsForChat: deps.evictLiveSessionsForChat,
      });
    case 'setNoMention':
      return performSetNoMention({ projectName: op.project, on: op.on });
    case 'setAutoCompact':
      return performSetAutoCompact({
        projectName: op.project,
        on: op.on,
        evictLiveSessionsForChat: deps.evictLiveSessionsForChat,
      });
  }
}

File: internal/admin/ops.go
package admin

// ops.go —— 管理面写操作层（对齐 TS admin/ops）。
// DM 卡片回调与 Web 控制台（未来）都只走这里写 projects.json——同一套校验/落盘/驱逐。
// 依赖注入：backendFor 创建后端、evict 驱逐活跃会话（orchestrator 进程内状态，必须在 bot 进程执行）。

import (
	"context"
	"fmt"
	"sync"
	"time"

	"github.com/modelzen/feishu-codex-bridge/internal/agent"
	"github.com/modelzen/feishu-codex-bridge/internal/project"
)

// OpKind 写操作类型。
type OpKind string

const (
	OpSwitchBackend     OpKind = "switchBackend"
	OpSetPermissionMode OpKind = "setPermissionMode"
	OpSetNoMention      OpKind = "setNoMention"
	OpSetAutoCompact    OpKind = "setAutoCompact"
)

// WriteOp Web/IPC 写操作的序列化形态。
type WriteOp struct {
	Kind      OpKind               `json:"kind"`
	Project   string               `json:"project"`
	Backend   string               `json:"backend,omitempty"`   // switchBackend
	Mode      agent.PermissionMode `json:"mode,omitempty"`      // setPermissionMode
	GuestMode agent.PermissionMode `json:"guestMode,omitempty"` // setPermissionMode
	Network   *bool                `json:"network,omitempty"`   // setPermissionMode
	On        bool                 `json:"on,omitempty"`        // setNoMention/setAutoCompact
}

// WriteOutcome perform* 的统一返回：ok 带写后回读的项目；!ok 带中文拒因。不抛错。
type WriteOutcome struct {
	Ok      bool
	Project *project.Project
	Reason  string
}

// WriteError 写操作被校验拒绝（HTTP 409 / IPC code 还原）。
type WriteError struct{ Reason string }

func (e *WriteError) Error() string { return e.Reason }

// BackendProbeRow 后端检测行（probeBackends 返回）。
type BackendProbeRow struct {
	ID             string
	Name           string
	Probe          *agent.BackendProbe
	SupportedModes []agent.PermissionMode
}

var allTiers = []agent.PermissionMode{agent.PermissionQA, agent.PermissionWrite, agent.PermissionFull}

// TierLabel 权限档中文标签（qa→项目内只读 / write→项目内读写 / full→完全访问）。
func TierLabel(m agent.PermissionMode) string {
	switch m {
	case agent.PermissionQA:
		return "项目内只读"
	case agent.PermissionWrite:
		return "项目内读写"
	case agent.PermissionFull:
		return "完全访问"
	}
	return string(m)
}

// ValidateBackendSwitch 项目后端切换的纯校验：
// ① 目标 id 在注册表；② doctor 探测通过；③ 项目两档权限都在目标后端支持面内。
// 全过返回空；否则返回中文原因。supportedModes nil = 全支持（codex）。
func ValidateBackendSwitch(target string, registered []string, proj project.Project, supportedModes []agent.PermissionMode, probe *agent.BackendProbe) string {
	found := false
	for _, id := range registered {
		if id == target {
			found = true
			break
		}
	}
	if !found {
		return fmt.Sprintf("未知后端「%s」（可用：%s）", target, joinIDs(registered))
	}
	if probe == nil || !probe.Ok {
		hint := "环境探测失败（未安装或未登录）"
		if probe != nil && probe.Hint != "" {
			hint = probe.Hint
		}
		return fmt.Sprintf("后端「%s」当前不可用：%s", target, hint)
	}
	if supportedModes != nil {
		mode := project.EffectiveMode(proj)
		guest := project.EffectiveGuestMode(proj)
		tiers := uniqueTiers(mode, guest)
		for _, t := range tiers {
			if !tierIn(supportedModes, t) {
				return fmt.Sprintf("该后端仅支持 %s 权限档，本项目当前为 %s —— 请先在「🔐 权限」把两档都调整到支持的档位再切换。",
					joinTiers(supportedModes), joinTiers(tiers))
			}
		}
	}
	return ""
}

// ProbeBackends 并行 doctor 探测全部后端（单超时，超时/抛错归一成 probe=nil）。
func ProbeBackends(ctx context.Context, backends []agent.AgentBackend, timeout time.Duration) []BackendProbeRow {
	if timeout <= 0 {
		timeout = 3 * time.Second
	}
	rows := make([]BackendProbeRow, len(backends))
	var wg sync.WaitGroup
	for i, be := range backends {
		i, be := i, be
		wg.Add(1)
		go func() {
			defer wg.Done()
			pctx, cancel := context.WithTimeout(ctx, timeout)
			defer cancel()
			probe := be.Doctor(pctx, true)
			rows[i] = BackendProbeRow{ID: be.ID(), Name: be.DisplayName(), Probe: &probe, SupportedModes: be.SupportedModes()}
		}()
	}
	wg.Wait()
	return rows
}

// Deps 写操作的进程内依赖（注入）。
type Deps struct {
	Store                    *project.Store
	BackendFor               func(id string) (agent.AgentBackend, error)
	EvictLiveSessionsForChat func(chatID string)
}

// RunWriteOp AdminWriteOp 分发 → 对应 perform*。
func RunWriteOp(ctx context.Context, op WriteOp, deps Deps) WriteOutcome {
	switch op.Kind {
	case OpSetNoMention:
		return PerformSetNoMention(deps.Store, op.Project, op.On)
	case OpSetAutoCompact:
		return PerformSetAutoCompact(deps.Store, op.Project, op.On, deps.EvictLiveSessionsForChat)
	case OpSetPermissionMode:
		return PerformSetPermissionMode(deps.Store, op.Project, op.Mode, op.GuestMode, op.Network, deps.EvictLiveSessionsForChat)
	case OpSwitchBackend:
		return PerformBackendSwitch(ctx, deps.Store, op.Project, op.Backend, deps.BackendFor)
	}
	return WriteOutcome{Ok: false, Reason: "未知写操作类型：" + string(op.Kind)}
}

// CreateWriteExecutor Web/IPC 入口：op → outcome，!ok 抛 WriteError。
func CreateWriteExecutor(deps Deps) func(context.Context, WriteOp) error {
	return func(ctx context.Context, op WriteOp) error {
		outcome := RunWriteOp(ctx, op, deps)
		if !outcome.Ok {
			return &WriteError{Reason: outcome.Reason}
		}
		return nil
	}
}

// PerformSetNoMention ✋ 免@ 开关（即时生效，无需驱逐）。
func PerformSetNoMention(store *project.Store, projectName string, on bool) WriteOutcome {
	p, err := store.GetByName(projectName)
	if err != nil || p == nil {
		return WriteOutcome{Ok: false, Reason: "项目「" + projectName + "」不存在"}
	}
	if err := store.Update(projectName, func(proj *project.Project) { proj.NoMention = boolPtr(on) }); err != nil {
		return WriteOutcome{Ok: false, Reason: err.Error()}
	}
	updated := *p
	updated.NoMention = boolPtr(on)
	return WriteOutcome{Ok: true, Project: &updated}
}

// PerformSetAutoCompact 🗜️ 自动压缩开关（落盘后驱逐让下一条消息重绑）。
func PerformSetAutoCompact(store *project.Store, projectName string, on bool, evict func(string)) WriteOutcome {
	p, err := store.GetByName(projectName)
	if err != nil || p == nil {
		return WriteOutcome{Ok: false, Reason: "项目「" + projectName + "」不存在"}
	}
	if err := store.Update(projectName, func(proj *project.Project) { proj.AutoCompact = boolPtr(on) }); err != nil {
		return WriteOutcome{Ok: false, Reason: err.Error()}
	}
	if evict != nil {
		evict(p.ChatID)
	}
	updated := *p
	updated.AutoCompact = boolPtr(on)
	return WriteOutcome{Ok: true, Project: &updated}
}

// PerformSetModelDefault 🤖 设置新话题默认模型/推理强度（不驱逐，只管新会话）。
func PerformSetModelDefault(store *project.Store, projectName, model string, effort agent.ReasoningEffort) WriteOutcome {
	p, err := store.GetByName(projectName)
	if err != nil || p == nil {
		return WriteOutcome{Ok: false, Reason: "项目「" + projectName + "」不存在"}
	}
	if err := store.Update(projectName, func(proj *project.Project) {
		proj.DefaultModel = model
		proj.DefaultEffort = effort
	}); err != nil {
		return WriteOutcome{Ok: false, Reason: err.Error()}
	}
	updated := *p
	updated.DefaultModel = model
	updated.DefaultEffort = effort
	return WriteOutcome{Ok: true, Project: &updated}
}

// PerformSetPermissionMode 🔐 权限档（落盘 mode/guestMode/network + 驱逐活跃会话）。
func PerformSetPermissionMode(store *project.Store, projectName string, mode, guestMode agent.PermissionMode, network *bool, evict func(string)) WriteOutcome {
	for _, v := range []agent.PermissionMode{mode, guestMode} {
		if v != "" && !tierIn(allTiers, v) {
			return WriteOutcome{Ok: false, Reason: "未知权限档「" + string(v) + "」"}
		}
	}
	p, err := store.GetByName(projectName)
	if err != nil || p == nil {
		return WriteOutcome{Ok: false, Reason: "项目「" + projectName + "」不存在"}
	}
	// 后端档位兼容守门：后端创建时固定，但权限档可改；改到该后端 supportedModes 外会卡死。
	if entry, ok := agent.CatalogByID(p.Backend); ok && len(entry.SupportedModes) > 0 {
		resMode := mode
		if resMode == "" {
			resMode = project.EffectiveMode(*p)
		}
		resGuest := guestMode
		if resGuest == "" {
			resGuest = project.EffectiveGuestMode(*p)
		}
		for _, t := range uniqueTiers(resMode, resGuest) {
			if !tierIn(entry.SupportedModes, t) {
				return WriteOutcome{Ok: false, Reason: fmt.Sprintf("项目的后端「%s」仅支持 %s 权限档，无法改到「%s」。", entry.DisplayName, joinTiers(entry.SupportedModes), TierLabel(t))}
			}
		}
	}
	if err := store.Update(projectName, func(proj *project.Project) {
		if mode != "" {
			proj.Mode = mode
		}
		if guestMode != "" {
			proj.GuestMode = guestMode
		}
		if network != nil {
			proj.Network = network
		}
	}); err != nil {
		return WriteOutcome{Ok: false, Reason: err.Error()}
	}
	if evict != nil {
		evict(p.ChatID)
	}
	updated := *p
	if mode != "" {
		updated.Mode = mode
	}
	if guestMode != "" {
		updated.GuestMode = guestMode
	}
	if network != nil {
		updated.Network = network
	}
	return WriteOutcome{Ok: true, Project: &updated}
}

// PerformBackendSwitch 🧠 切换项目后端（写盘前再 doctor force 探一次；不驱逐活跃会话）。
func PerformBackendSwitch(ctx context.Context, store *project.Store, projectName, target string, backendFor func(string) (agent.AgentBackend, error)) WriteOutcome {
	p, err := store.GetByName(projectName)
	if err != nil || p == nil {
		return WriteOutcome{Ok: false, Reason: "项目「" + projectName + "」不存在"}
	}
	// 后端运行时固定（仅 legacy 一次性落地 + 同值 no-op）。
	if p.Backend != "" && p.Backend != target {
		return WriteOutcome{Ok: false, Reason: "该项目的后端已在创建时选定，运行时固定、不支持切换。如需更改，请删除该项目后用新后端重新创建。"}
	}
	registered := agent.CatalogBackendIDs()
	be, err := backendFor(target)
	if err != nil {
		return WriteOutcome{Ok: false, Reason: ValidateBackendSwitch(target, registered, *p, nil, nil)}
	}
	probe := be.Doctor(ctx, true)
	supported := be.SupportedModes()
	if reason := ValidateBackendSwitch(target, registered, *p, supported, &probe); reason != "" {
		return WriteOutcome{Ok: false, Reason: reason}
	}
	if err := store.Update(projectName, func(proj *project.Project) { proj.Backend = target }); err != nil {
		return WriteOutcome{Ok: false, Reason: err.Error()}
	}
	updated := *p
	updated.Backend = target
	return WriteOutcome{Ok: true, Project: &updated}
}

// ── 小工具 ───────────────────────────────────────────────────────

func boolPtr(b bool) *bool { return &b }

func tierIn(modes []agent.PermissionMode, m agent.PermissionMode) bool {
	for _, x := range modes {
		if x == m {
			return true
		}
	}
	return false
}

func uniqueTiers(tiers ...agent.PermissionMode) []agent.PermissionMode {
	seen := map[agent.PermissionMode]bool{}
	out := []agent.PermissionMode{}
	for _, t := range tiers {
		if t != "" && !seen[t] {
			seen[t] = true
			out = append(out, t)
		}
	}
	return out
}

func joinTiers(tiers []agent.PermissionMode) string {
	labels := make([]string, 0, len(tiers))
	for _, t := range tiers {
		labels = append(labels, TierLabel(t))
	}
	return joinStr(labels, " / ")
}

func joinIDs(ids []string) string { return joinStr(ids, "、") }

func joinStr(ss []string, sep string) string {
	if len(ss) == 0 {
		return ""
	}
	out := ss[0]
	for _, s := range ss[1:] {
		out += sep + s
	}
	return out
}
